### PR TITLE
fix(Task.all): fixes Task.all when called with an empty array (closes #5)

### LINKED ISSUE #5: `Task.all` never resolves if it's called with an empty array
The following `Task` will never be _resolved_ nor _rejected_:

```typescript
Task.all([] as Task<number, never>[]);
```

It should be _resolved_ with an _empty array_.

### PR DESCRIPTION
**Pull request by dggluz**
# Description

Now `Task.all([])`, returns a `Task` _resolved_ with an _empty array_.

I also added the _overload_ to support the case when the _input array_ is `never[]` (the typechecker didn't complain before, but could not infer correctly the resulting `Task`).

# Related issues

Closes #5.

### CODE REVIEW

**Comment by hrajchert** [comment]
I'm not sure about the never[] overload... why is this necessary?

**Comment by dggluz** [comment]
That is because of the most generic typing overload (`function all <T, E> (tasks: Array<Task<T, E>>): Task<T[], E>`) that was allowing empty arrays but couldn't infer any resolution nor rejection types for them:

task-all-empty-array-error

I think that is better typed with the `never[]` case.

**Comment by dggluz** [comment]
I think another option is to default the generics to never in the most general case. It would be:
```typescript
function all <T = never, E = never> (tasks: Array<Task<T, E>>): Task<T[], E>
```

I don't have a preference for any of those options.

**Comment by hrajchert** [comment]
Well, once we upgrade to 3.x we could use the unknown type. Of those two options I would prefer the second one, but even better is not type it.

When you do
```
let a1 = [1, 2, 3]
a1.push(4);
```
you don't specify the type because typescript can infer it. But if you initialize with an empty array you normally specify the type beforehand.

```
let a2: number[] = [];
a2.push(1)
```

Having an array of never is not useful. If you don't want to get the `{}` type infered and prefer never or number you can always specify it

```
Task.all<never>([])
```

In conclusion, I would remove the never part.

**Comment by dggluz** [comment]
Ok, I removed that! I think the PR is ready to be merged!

**Review by hrajchert — APPROVED** [review]
